dbms/src/DataStreams/WindowBlockInputStream.cpp

```diff
@@ -27,6 +37,119 @@ extern const int BAD_ARGUMENTS;
 extern const int NOT_IMPLEMENTED;
 } // namespace ErrorCodes
 
+namespace
+{
+template <typename T>
+consteval bool checkIfSimpleNumericType()
+{
+    return std::is_integral_v<T> || std::is_floating_point_v<T>;
+}
+
+template <typename T>
+consteval bool checkIfDecimalFieldType()
+{
+    return std::is_same_v<T, DecimalField<Decimal32>> || std::is_same_v<T, DecimalField<Decimal64>> || std::is_same_v<T, DecimalField<Decimal128>> || std::is_same_v<T, DecimalField<Decimal256>>;
+}
+
+// When T is Decimal, we should convert it to DecimalField type
+// as we need scale value when executing the comparison operation.
+template <typename T>
+struct ActualCmpDataType
+{
+    using Type = std::conditional_t<checkIfSimpleNumericType<T>(), T, DecimalField<T>>;
+};
+
+template <typename T>
+typename ActualCmpDataType<T>::Type getValue(const ColumnPtr & col_ptr, size_t idx)
+{
+    return (*col_ptr)[idx].get<typename ActualCmpDataType<T>::Type>();
+}
+
+template <typename T, typename U, bool is_start, bool is_desc>
+bool isInRangeCommonImpl(T current_row_aux_value, U cursor_value)
+{
+    if constexpr ((is_start && is_desc) || (!is_start && !is_desc))
+        return cursor_value <= current_row_aux_value;
+    else
+        return cursor_value >= current_row_aux_value;
+}
+
+template <typename T, typename U, bool is_start, bool is_desc>
+bool isInRangeIntImpl(T current_row_aux_value, U cursor_value)
+{
+    return isInRangeCommonImpl<T, U, is_start, is_desc>(current_row_aux_value, cursor_value);
+}
+
+template <typename AuxColType, typename OrderByColType, bool is_start, bool is_desc>
+bool isInRangeDecimalImpl(AuxColType current_row_aux_value, OrderByColType cursor_value)
+{
+    if constexpr (checkIfDecimalFieldType<AuxColType>() && checkIfDecimalFieldType<OrderByColType>())
+        return isInRangeCommonImpl<AuxColType, OrderByColType, is_start, is_desc>(current_row_aux_value, cursor_value);
+
+    // When comparing Decimal and Int, we convert them to the Float first.
+    Float64 current_row_aux_value_float64;
```

Comment:
> Both `Decimal` and `Int` are acurate type, why convert them to in-acurate type before compare? Is MySQL/TiDB use the same compare stragety?

Because we need to reserve the fraction part of decimal before comparing and the accuracy loss of `Decimal` will not affect the result I think.

Reply:
I think you should first do some investigations on how TiFlash compare decimal value and int value, the accuracy loss is possible because double type only has 15 to 17 significant decimal digits precision, for example:
```
mysql> select value, cast(value as double), cast(value as double) + 1, cast(value as double) > cast(value as double) + 1, cast(value as double) < cast(value as double) + 1, cast(value as double) = cast(value as double) + 1 from t;
+---------------------+-----------------------+---------------------------+---------------------------------------------------+---------------------------------------------------+---------------------------------------------------+
| value               | cast(value as double) | cast(value as double) + 1 | cast(value as double) > cast(value as double) + 1 | cast(value as double) < cast(value as double) + 1 | cast(value as double) = cast(value as double) + 1 |
+---------------------+-----------------------+---------------------------+---------------------------------------------------+---------------------------------------------------+---------------------------------------------------+
| 9223372036854775801 |  9.223372036854776e18 |      9.223372036854776e18 |                                                 0 |                                                 0 |                                                 1 |
+---------------------+-----------------------+---------------------------+---------------------------------------------------+---------------------------------------------------+---------------------------------------------------+
```

Reply:
> I think you should first do some investigations on how TiFlash compare decimal value and int value, the accuracy loss is possible because double type only has 15 to 17 significant decimal digits precision, for example:
> 
> ```
> mysql> select value, cast(value as double), cast(value as double) + 1, cast(value as double) > cast(value as double) + 1, cast(value as double) < cast(value as double) + 1, cast(value as double) = cast(value as double) + 1 from t;
> +---------------------+-----------------------+---------------------------+---------------------------------------------------+---------------------------------------------------+---------------------------------------------------+
> | value               | cast(value as double) | cast(value as double) + 1 | cast(value as double) > cast(value as double) + 1 | cast(value as double) < cast(value as double) + 1 | cast(value as double) = cast(value as double) + 1 |
> +---------------------+-----------------------+---------------------------+---------------------------------------------------+---------------------------------------------------+---------------------------------------------------+
> | 9223372036854775801 |  9.223372036854776e18 |      9.223372036854776e18 |                                                 0 |                                                 0 |                                                 1 |
> +---------------------+-----------------------+---------------------------+---------------------------------------------------+---------------------------------------------------+---------------------------------------------------+
> ```

fixed